### PR TITLE
3 user auth

### PR DESCRIPTION
**Pull request by afg419**
This PR builds up a Bcrypt backed auth system using the comeonin library in the backend.  It integrates this with a burgeoning front end in React, and the previous websocket token increment functionality.  It is hardly a pretty child, but it is mine and I love it all the same.

### CODE REVIEW

File: count_server/test/models/user_test.exs [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+  @valid_attrs %{username: "axeface", password: "password"}
+  @invalid_attrs %{}
+
+    test "changeset with valid attributes" do
+    changeset = User.changeset(%User{}, @valid_attrs)
+    assert changeset.valid?
+  end
+
+  test "changeset with invalid attributes" do
+    changeset = User.changeset(%User{}, %{username: "axeface"})
```
Comment (hectorhuertas) [inline count_server/test/models/user_test.exs:16]: axeface should not need a password. Should be ALWAYS valid

File: count_server/web/router.ex [hunk 2]
```diff
@@ -16,11 +16,17 @@ defmodule CountServer.Router do
...
 
-    get "/", CounterController, :index
+    # get "/", CounterController, :index
+    get "/", HomeController, :index
   end
 
   # Other scopes may use custom stacks.
-  # scope "/api", CountServer do
-  #   pipe_through :api
-  # end
```
Comment (hectorhuertas) [inline count_server/web/router.ex:25]: Get rid of these comments?, you won't need them in the future

File: count_server/web/router.ex [hunk 3]
```diff
@@ -16,11 +16,17 @@ defmodule CountServer.Router do
...
+  scope "/api/v1", CountServer do
```
Comment (hectorhuertas) [inline count_server/web/router.ex:24]: This is something i did too, and I don't know if its the best choice.
The url is scoped by /api/v1, but not the path helpers or the module names.
Will this be a problem in a future v2 of the api? I have no idea. 
Will this create conflicts if you need a different SessionController for the browser and the api pipelines? Probably

As I said I don't know if it is going to affect, but in rails it probably will

File: count_server/web/views/session_view.ex [hunk 4]
```diff
@@ -0,0 +1,15 @@
+defmodule CountServer.SessionView do
+  use CountServer.Web, :view
+
+  def render("new.json", %{reply: reply}) do
+    reply
+  end
+
```
Comment (hectorhuertas) [inline count_server/web/views/session_view.ex:7]: I think you can write this in one line:
def render("new.json", %{reply: reply}), do: reply

File: count_server/web/views/session_view.ex [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+  def render("create.json", %{reply: reply}) do
+    reply
+  end
+
+  def render("destroy.json", %{reply: reply}) do
+    reply
+  end
+end
```
Comment (hectorhuertas) [inline count_server/web/views/session_view.ex:15]: If you are returning 'reply' no matter the action, maybe you don't need to be pattern matching for the action.
You whole controller could be:

```
defmodule CountServer.SessionView do
  use CountServer.Web, :view

  def render(_action, %{reply: reply}), do: reply
end
```

**Comment by bethsecor** [comment]
You should probably squash these commits!

File: count_server/mix.exs [hunk 6]
```diff
@@ -14,14 +14,33 @@ defmodule CountServer.Mixfile do
   end
 
   # Configuration for the OTP application.
-  #
+
```
Comment (bethsecor) [inline count_server/mix.exs:17]: Do you still need these comments?

File: count_server/web/channels/counter_channel.ex [hunk 7]
```diff
@@ -1,12 +1,35 @@
...
-  def handle_in("count_up", %{"body" => body}, socket) do
-    broadcast! socket, "count_up", %{body: body}
+  def handle_in("count_up", _params, socket) do
+    counter = Repo.get!(Counter, 1)
+    new_value = counter.main + 1
+
+    changeset = Counter.changeset(counter, %{main: new_value})
+
+
+    # post = MyRepo.get!(Post, 42)
```
Comment (bethsecor) [inline count_server/web/channels/counter_channel.ex:19]: Commented code!!!

File: count_server/web/controllers/session_controller.ex [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+
+  alias CountServer.Repo
+  alias CountServer.User
+  alias CountServer.AppState
+
+  alias Comeonin.Bcrypt
+
+  def new(conn, %{"password" => password, "username" => username}) do
+    changeset = %User{}
+    |> User.changeset(%{password: Bcrypt.hashpwsalt(password), username: username})
```
Comment (bethsecor) [inline count_server/web/controllers/session_controller.ex:12]: This is cool.

File: count_server/web/static/js/app.js [hunk 9]
```diff
@@ -19,3 +19,164 @@ import "phoenix_html"
...
+        <div>  |-|  </div>
+        <div>  |v|  </div>
+        {this.renderButtons()}
+        <div> {this.state.message} </div>
+      </div>
+    );
+  }
+});
+
+var Game = React.createClass({
```
Comment (bethsecor) [inline count_server/web/static/js/app.js:154]: Should you put this class in a separate file?

File: count_server/web/router.ex (hunk 3 above)
Comment (afg419) [inline count_server/web/router.ex:24]: hmm.  Well don't think I'll need a sessionController for the browser pipeline as I'm building my front end in React, but really good point.  I don't know either.

File: count_server/web/views/session_view.ex (hunk 5 above)
Comment (afg419) [inline count_server/web/views/session_view.ex:15]: Nice.

File: count_server/mix.exs (hunk 6 above)
Comment (afg419) [inline count_server/mix.exs:17]: lol no, they came with phoenix new.  Deleting them is kind of like deleting prepackaged comments in your gemfile.  I got rid of them

File: count_server/web/static/js/app.js (hunk 9 above)
Comment (afg419) [inline count_server/web/static/js/app.js:154]: absolutely.

**Pull request merged**